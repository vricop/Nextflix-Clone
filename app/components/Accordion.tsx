import HTMLReactParser from 'html-react-parser'
import { Plus } from 'iconoir-react'
import { MouseEvent, ReactNode } from 'react'
import styled from 'styled-components'

const Container = styled.div`
  width: 75%;
  max-width: 815px;
  margin-inline: auto;
  display: grid;
  grid-auto-flow: row;
  row-gap: var(--size-2);
  font-size: var(--text-2xl);
`

const Details = styled.details`
  display: grid;
  grid-auto-flow: row;
  row-gap: var(--size-2);

  &[open]:not(.js) > summary + *,
  &.open > summary + * {
    max-height: calc(100vh - var(--size-4) * 2);
    padding-block: var(--size-4);
  }

  &[open] > summary > svg {
    transform: rotate(-45deg);
  }
`

const Summary = styled.summary`
  background-color: var(--color-gray-800);
  padding: var(--size-5);
  display: grid;
  grid-auto-flow: column;
  justify-content: space-between;
  column-gap: var(--size-4);
  cursor: pointer;

  & > svg {
    transition: transform 400ms cubic-bezier(0.45, 0.05, 0.55, 0.95);
    transform: rotate(0);
  }

  & + * {
    background-color: var(--color-gray-800);
    transition: 400ms cubic-bezier(0.45, 0.05, 0.55, 0.95);
    overflow: hidden;
    max-height: 0vh;
    padding-inline: var(--size-4);
    margin-top: var(--size-px);
  }
`

export function Accordion({ children }: { children: ReactNode }) {
  return <Container>{children}</Container>
}

function collapseDetails(openDetails: HTMLDetailsElement) {
  const detailsCollection = openDetails
    .closest('div')!
    .querySelectorAll('details')

  detailsCollection.forEach(item => {
    if (item === openDetails) return

    item.classList.remove('open')

    timer2 = setTimeout(() => {
      item.open = false
    }, 400)
  })
}

let timer1: ReturnType<typeof setTimeout>
let timer2: ReturnType<typeof setTimeout>
function handleClick(event: MouseEvent<HTMLDetailsElement>) {
  event.preventDefault()
  const { currentTarget: details } = event

  clearTimeout(timer1)
  clearTimeout(timer2)

  details.classList.toggle('open')
  if (!details.classList.contains('js')) details.classList.add('js')

  if (!details.open) {
    details.open = true
    collapseDetails(details)
    return
  }

  timer1 = setTimeout(() => {
    details.open = false
  }, 400)
}

const AccordionItem = ({
  summary,
  details,
}: {
  summary: string
  details: string
}) => (
  <Details onClick={handleClick}>
    <Summary>
      {summary}
      <Plus />
    </Summary>
    <div>{HTMLReactParser(details)}</div>
  </Details>
)

Accordion.Item = AccordionItem
