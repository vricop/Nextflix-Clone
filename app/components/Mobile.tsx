import styled from 'styled-components'

const Container = styled.div`
  display: grid;
  place-items: center;
`

const Phone = styled.img`
  grid-area: 1/1;
`

const DownloadItemCard = styled.div`
  align-items: center;
  align-self: end;
  background-color: var(--color-black);
  border-radius: var(--size-3);
  border: var(--size-0-5) solid var(--color-gray-700);
  column-gap: var(--size-4);
  display: grid;
  grid-area: 1/1;
  grid-template-columns: auto 1fr auto;
  padding: var(--size-2) var(--size-3);
  transform: translateY(-22%);
  width: 60%;
`

const Poster = styled.img`
  width: auto;
  height: var(--size-20);
  aspect-ratio: 150/210;
`

const Title = styled.h3`
  font-size: var(--font-base);
  line-height: var(--leading-tight);
`

const Description = styled.p`
  font-size: var(--text-sm);
  color: var(--color-blue);
  line-height: var(--leading-tight);
`

const Gif = styled.img`
  height: var(--size-12);
`

export function Mobile() {
  return (
    <Container>
      <Phone src="/images/mobile-0819.jpg" alt="Mobile - Stranger Things" />
      <DownloadItemCard>
        <Poster
          width={150}
          height={210}
          src="/images/boxshot.png"
          alt="Stranger Things Poster"
        />
        <div>
          <Title>Stranger Things</Title>
          <Description>Downloading...</Description>
        </div>
        <Gif src="/images/download-icon.gif" alt="" />
      </DownloadItemCard>
    </Container>
  )
}
