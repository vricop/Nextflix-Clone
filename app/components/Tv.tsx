import styled from 'styled-components'

const Wrapper = styled.div`
  display: grid;
  place-items: center;
`

const Frame = styled.img`
  z-index: 1;
  grid-area: 1/1;
`

const Video = styled.video`
  width: 100%;
  height: 100%;
  max-width: 73%;
  aspect-ratio: 1.333;
  top: -1.8%;
  position: relative;
  left: -0.3%;
  grid-area: 1/1;
`

export const Tv = () => (
  <Wrapper>
    <Frame src="/images/tv.png" alt="TV" />
    <Video autoPlay muted loop playsInline>
      <source src="/videos/video-tv-0819.m4v.mp4" type="video/mp4" />
    </Video>
  </Wrapper>
)
