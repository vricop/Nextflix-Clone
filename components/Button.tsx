import { ReactElement, ReactNode } from 'react'

interface ButtonProps {
  children: ReactNode
}

export function Button({ children }: ButtonProps): ReactElement {
  return <button type="button">{children}</button>
}
