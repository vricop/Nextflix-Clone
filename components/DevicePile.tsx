import styled from 'styled-components'

const Container = styled.div`
  display: grid;
  place-items: center;
`

const Devices = styled.img`
  grid-area: 1/1;
  z-index: 1;
`

const Video = styled.video`
  grid-area: 1/1;
  width: 100%;
  height: 100%;
  max-width: 63%;
  max-height: 47%;
  transform: translate(0%, -34%);
`

export function DevicePile() {
  return (
    <Container>
      <Devices src="/images/device-pile.png" alt="" />
      <Video autoPlay playsInline muted loop>
        <source src="/videos/video-devices.mp4" type="video/mp4" />
      </Video>
    </Container>
  )
}
