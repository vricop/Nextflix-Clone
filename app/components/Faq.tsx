import { useTranslation } from 'react-i18next'
import styled from 'styled-components'
import { v4 as uuid } from 'uuid'
import { Accordion } from './Accordion'
import { Cta } from './common/Cta'
import { EmailBox } from './common/EmailBox'
import { HomeContainer } from './common/HomeContainer'
import { HomeSectionTitle } from './common/HomeSectionTitle'

const Title = styled(HomeSectionTitle)`
  text-align: center;
`

const FaqContainer = styled(HomeContainer)`
  display: grid;
  grid-auto-flow: row;
  justify-items: center;
  row-gap: ${({ theme }) => theme.size._6};

  @media (min-width: ${({ theme }) => theme.breakpoints.sm}) {
    row-gap: ${({ theme }) => theme.size._8};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.lg}) {
    row-gap: ${({ theme }) => theme.size._13};
  }
`

const FaqAccordion = styled(Accordion)`
  width: 100%;

  @media (min-width: ${({ theme }) => theme.breakpoints.sm}) {
    width: 90%;
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    width: 75%;
    max-width: 815px;
  }
`

const CTAWrapper = styled.div`
  text-align: center;
  display: grid;
  justify-items: center;
  row-gap: ${({ theme }) => theme.size._5};
`

export function Faq() {
  const { t } = useTranslation('home')

  return (
    <FaqContainer role="">
      <Title>{t('faqTitle')}</Title>
      <FaqAccordion>
        {t('faq').map(details => (
          <FaqAccordion.Item key={uuid()} {...details} />
        ))}
      </FaqAccordion>
      <CTAWrapper>
        <Cta />
        <EmailBox />
      </CTAWrapper>
    </FaqContainer>
  )
}
