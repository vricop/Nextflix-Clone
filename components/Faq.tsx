import { useTranslation } from 'next-i18next'
import styled from 'styled-components'
import { v4 as uuid } from 'uuid'
import { Accordion, Item } from './Accordion'
import { Cta } from '@components/Cta'
import { EmailBox } from '@components/EmailBox'
import { HomeContainer } from '@components/HomeContainer'
import { HomeSectionTitle } from '@components/HomeSectionTitle'

const Title = styled(HomeSectionTitle)`
  text-align: center;
`

const Container = styled(HomeContainer)`
  display: grid;
  grid-auto-flow: row;
  justify-items: center;
  row-gap: ${({ theme }) => theme.size._6};

  @media (min-width: ${({ theme }) => theme.breakpoints.sm}) {
    row-gap: ${({ theme }) => theme.size._8};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.lg}) {
    row-gap: ${({ theme }) => theme.size._13};
  }
`

const FaqAccordion = styled(Accordion)`
  width: 100%;

  @media (min-width: ${({ theme }) => theme.breakpoints.sm}) {
    width: 90%;
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    width: 75%;
    max-width: 815px;
  }
`

FaqAccordion.displayName = 'Faq__Accordion'

const CtaWrapper = styled.div`
  text-align: center;
  display: grid;
  justify-items: center;
  row-gap: ${({ theme }) => theme.size._5};
`

export function Faq() {
  const { t } = useTranslation('home')

  return (
    <Container role="">
      <Title>{t('faqTitle')}</Title>
      <FaqAccordion>
        {t('faq').map(details => (
          <Item key={uuid()} {...details} />
        ))}
      </FaqAccordion>
      <CtaWrapper>
        <Cta />
        <EmailBox />
      </CtaWrapper>
    </Container>
  )
}
