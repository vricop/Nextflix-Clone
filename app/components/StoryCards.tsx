import { ReactElement } from 'react'
import { useTranslation } from 'react-i18next'
import { v4 as uuid } from 'uuid'
import { DevicePile } from './DevicePile'
import { Mobile } from './Mobile'
import { StoryCard } from './StoryCard'
import { StoryImage } from './StoryImage'
import { Tv } from './Tv'

const media: { [index: number]: ReactElement } = {
  0: <Tv />,
  1: <Mobile />,
  2: <DevicePile />,
  3: <StoryImage />,
}

export default function StoryCards() {
  const { t } = useTranslation('home')
  const storyCards = t('storyCards')

  return (
    <>
      {storyCards.map((storyCard, index) => (
        <StoryCard
          key={uuid()}
          flipped={index % 2 !== 0}
          title={storyCard.title}
          tagline={storyCard.tagline}>
          {media[index]}
        </StoryCard>
      ))}
    </>
  )
}
