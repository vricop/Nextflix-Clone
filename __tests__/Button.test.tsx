import { Button } from '@/components/Button'
import { render, screen } from '@testing-library/react'

it('renders a button', () => {
  const { container, debug } = render(<Button>Cool button</Button>)
  debug(container)
  expect(container).toMatchSnapshot()
})
