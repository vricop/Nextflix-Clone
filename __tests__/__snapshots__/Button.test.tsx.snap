// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a button 1`] = `
<div>
  <button
    type="button"
  >
    Cool button
  </button>
</div>
`;
